src/script.rs

```diff
@@ -0,0 +1,642 @@
+                    let read_le = |script: &mut &[u8], needed_bytes: usize| {
+                        if script.len() < needed_bytes {
+                            Err(ScriptError::ReadError {
+                                expected_bytes: needed_bytes,
+                                available_bytes: script.len(),
+                            })
+                        } else {
+                            let mut size = 0;
+                            for i in (0..needed_bytes).rev() {
+                                size <<= 8;
+                                size |= usize::from(script[i]);
+                            }
+                            *script = &script[needed_bytes..];
+                            Ok(size)
+                        }
+                    };
+
+                    let size = match pv {
+                        OP_PUSHDATA1 => read_le(script, 1),
+                        OP_PUSHDATA2 => read_le(script, 2),
+                        OP_PUSHDATA4 => read_le(script, 4),
+                        _ => unreachable!(),
+                    }?;
+
+                    if script.len() < size {
+                        return Err(ScriptError::ReadError {
+                            expected_bytes: size,
+                            available_bytes: script.len(),
+                        });
+                    }
+
+                    buffer.extend(&script[0..size]);
+                    *script = &script[size..];
+
+                    leading_byte
+                }
+                // OP_0/OP_FALSE doesn't actually push a constant 0 onto the stack but
+                // pushes an empty array. (Thus we leave the buffer truncated to 0 length)
+                OP_0 => leading_byte,
+                PushdataBytelength(size_byte) => {
+                    let size = size_byte.into();
+
+                    if script.len() < size {
+                        return Err(ScriptError::ReadError {
+                            expected_bytes: size,
+                            available_bytes: script.len(),
+                        });
+                    }
+
+                    buffer.extend(&script[0..size]);
+                    *script = &script[size..];
+
+                    leading_byte
+                }
+                _ => leading_byte,
+            },
+            _ => leading_byte,
+        })
+    }
+
+    /** Encode/decode small integers: */
+    pub fn decode_op_n(opcode: PushValue) -> u32 {
+        if opcode == OP_0 {
+            return 0;
+        }
+        assert!(opcode >= OP_1 && opcode <= OP_16);
+        (u8::from(opcode) - (u8::from(OP_1) - 1)).into()
+    }
+
+    /// Pre-version-0.6, Bitcoin always counted CHECKMULTISIGs
+    /// as 20 sigops. With pay-to-script-hash, that changed:
+    /// CHECKMULTISIGs serialized in script_sigs are
+    /// counted more accurately, assuming they are of the form
+    ///  ... OP_N CHECKMULTISIG ...
```

Comment:
Blech, we should get rid of this nonsense.

Reply:
This is meant as a suggestion for the future, right? Because this would lead to a consensus fork, since the C++ version would reject blocks that Rust sees as valid.

Reply:
Yes, that was my intention. It would have to be done carefully, but I believe it's not consensus.